content/posts/wasm-as-a-platform-for-abstraction.md

```diff
@@ -0,0 +1,1390 @@
+
+#![allow(bad_style, dead_code)]
+
+// /src/libc/unix/linux_like/linux/gnu/b64/x86_64/mod.rs
+pub type c_char = i8;
+pub type wchar_t = i32;
+
+// /src/libc/unix/linux_like/linux/gnu/b64/x86_64/not_x32.rs
+pub type c_long = i64;
+pub type c_ulong = u64;
+
+// src/libc/unix/mod.rs
+pub type c_schar = i8;
+pub type c_uchar = u8;
+pub type c_short = i16;
+pub type c_ushort = u16;
+pub type c_int = i32;
+pub type c_uint = u32;
+pub type c_float = f32;
+pub type c_double = f64;
+pub type c_longlong = i64;
+pub type c_ulonglong = u64;
+pub type intmax_t = i64;
+pub type uintmax_t = u64;
+pub type size_t = usize;
+pub type ptrdiff_t = isize;
+pub type intptr_t = isize;
+pub type uintptr_t = usize;
+pub type ssize_t = isize;
+```
+
+We can now generate declarations for `intrinsics.h`. This will
+be analogous to [`std::intrinsics`][std-intrinsics] in Rust's standard library.
+
+```console
+$ cp ../intrinsics.h .
+$ bindgen intrinsics.h \
+    --whitelist-type 'wasm_.*' \
+    --whitelist-function 'wasm_.*' \
+    --output src/intrinsics.rs \
+    --use-core \
+    --ctypes-prefix crate::ctypes \
+     --raw-line '#![allow(bad_style, dead_code)]'
+$ tail src/intrinsics.rs
+extern "C" {
+    /// Write to a globally defined integer variable.
+    ///
+    /// This may fail if the variable already exists and has a different type.
+    pub fn wasm_variable_write_int(
+        name: *const ::std::os::raw::c_char,
+        name_len: ::std::os::raw::c_int,
+        value: i32,
+    ) -> wasm_result_t;
+}
+```
+
+Our `lib.rs` needs to be updated to use `intrinsics`.
+
+```rust
+// std/lib.rs
+
+//! The standard library, providing host bindings and abstractions.
+
+// we are the standard library.
+#![no_std]
+
+pub mod intrinsics;
+```
+
+While it's still quite small at the moment, as we gain more experience using
+this system we'll be able to move commonly-used elements into the standard
+library to provide a more *batteries included* feel.
+
+Seeing as the end goal is for users to write programs for our controller in
+any language, not just Rust, this may eventually require tools like
+[*Interface Types*][interface-types].
+
+Now we have a standard library we can rewrite our previous `example-program.rs`.
+
+```console
+$ cd ../examples/wasm-programs
+$ rm example-program.rs
+$ cargo new example-program
+$ cd example-program
+```
+
+We need to add our standard library as a dependency.
+
+```console
+$ cargo add ../../../std
+    Updating 'https://github.com/rust-lang/crates.io-index' index
+      Adding wasm-std (unknown version) to dependencies
+```
+
+Because this crate is being compiled to WASM we'll need to make sure it is
+compiled using the `cdylib` crate type.
+
+```toml
+# examples/wasm-programs/example-program/Cargo.toml
+
+[package]
+name = "example-program"
+version = "0.1.0"
+authors = ["Michael Bryan <michaelfbryan@gmail.com>"]
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+wasm-std = { path = "../../../../iec-std/" }
```

Comment:
I guess "iec" is not introduced earlier in the article?

Reply:
Oops. I tried to deliberately remove all references to `iec` and `rustmatic`. Looks like I missed one.